x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessHeaders.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.xpack.core.security.authc.RemoteAccessAuthentication;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public final class RemoteAccessHeaders {
+
+    public static final String REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY = "_remote_access_cluster_credential";
+    private final EncodedApiKeyWithPrefix clusterCredentials;
+    private final RemoteAccessAuthentication remoteAccessAuthentication;
+
+    public RemoteAccessHeaders(String encodedCredentials, RemoteAccessAuthentication remoteAccessAuthentication) {
+        this(new EncodedApiKeyWithPrefix(encodedCredentials), remoteAccessAuthentication);
+    }
+
+    private RemoteAccessHeaders(EncodedApiKeyWithPrefix clusterCredentials, RemoteAccessAuthentication remoteAccessAuthentication) {
```

Comment:
I considered `ApiKeyCredentials` here instead but decided against because of the following scenario:

If we ever update the format of `ApiKeyCredentials` and we have a new FC with the new format and an older QC without the format change, the QC would re-serialize the API key according to the old format, which wouldn't be parsed by the FC without extra BWC handling.

If we instead just pass the encoded API key through from the QC to the FC, there is no BWC concern.

Reply:
I agree that we should use the encoded string here instead of `ApiKeyCredentials`. But the `EncodedApiKeyWithPrefix` class feels a bit bloated and unncessary to me. I think it is fine for `RemoteAccessHeader` to house these logic. Also, the `EncodedApiKeyWithPrefix` caches the `ApiKeyCredentials` parsing resul so that it does not have to parse twice. But in the main code, we only parse it once anyway. So the saving isn't real. The usage of `RemoteAccessHeader` is scoped. So even if we need `ApiKeyCredentials` in more than 1 place, the caller can just save it to a variable and pass that variable around. If we want to be extra cautious, we can comment the parsing method to say "parsing result is not cached".

I'd suggest we drop `EncodedApiKeyWithPrefix` for something like: 

```java
class RemoteAccessHeaders {
    String encodedCredentials;
    RemoteAccessAuthentication remoteAccessAuthentication;

    AuthenticationToken clusterCredentials() { 
         // in future, we maybe able to delegate this to a common credential extrator
    }
    RemoteAccessAuthentication  remoteAccessAuthentication() {}
    writeToContext(...) {}
    readFromContext(...) {}
}
```

Reply:
`EncodedApiKeyWithPrefix` is indeed a bit verbose but I have a slight preference to keep it:

If we remove it, and we don't keep a nulleable `ApiKeyCredentials` field on `RemoteAccessHeaders`, we have to do an awkward and redundant dance around stripping the ApiKey prefix just to add it back again in `clusterCredentials` (just to strip it, yet again, in the underlying call to extractFromHeader). This is by no means a performance problem, just doesn't feel very clean. 

More concretely, we'd have:

```java
readFromContext(...) {
  var encoded = stripPrefix(header)
  ...
}

AuthenticationToken clusterCredentials() {
  // add back the header we just removed
  extractCredentials(withHeader(encoded))
  ...  
}

// duplicate validation logic that prefix is present, which we already have in `extractCredentials`
stripPrefix() {
}
```

Furthermore, in the current approach `readFromContext` throws if we have an invalid credential; if we move API key parsing to `clusterCredentials` we end up creating the headers object and only throwing when we try to access one of its fields, which seems less friendly than failing at creation time. 

I think `EncodedApiKeyWithPrefix` encapsulates this reasonably well -- I would be alright with inlining it but would prefer to at least include a nullable ApiKeyCredentials field in `RemoteAccessHeaders` in that case, to avoid the redundant prefix-stripping and fail parsing at creation time, instead of when we call `clusterCredentials`.

LMWYT!

Reply:
> we end up creating the headers object and only throwing when we try to access one of its fields, which seems less friendly than failing at creation time.

Isn't this exactly why we want by sending the encoded string of cluster credentials across instead of `ApiKeyCrentials`? When we create `RemoteAccessHeaders` on the QC side, we don't want to parse the string (because parsing can even fail due to version difference) but instead just sending it over and let FC parse it. This means we don't want to parse it at creation time, but rather when we actually need it. 

Maybe I am missing something, but I also don't know why we need actively `stripPrefix`. The `RemoteClusterAuthorizationResolver`returns a cluster credentials with the prefix already because prefix is part of the authorization information. `RemoteAccessHeaders` just keep a reference to this prefixed string. Then parse it for ApiKeyCredentials when needed. Ok, our difference probably still goes back to whether we want to parse at creation time. I don't think we need to. 

This constructor really bothers me 
```
        private EncodedApiKeyWithPrefix(String encodedApiKey, @Nullable ApiKeyService.ApiKeyCredentials apiKeyCredentials) {
            this.apiKeyCredentials = apiKeyCredentials;
            this.encodedApiKey = encodedApiKey;
        }
```
It has two fields with duplicated information without anything guarantee them to be consistent. We can add consistency check. But it overall feels better to not have it at all.

Reply:
We had a short sync: I've updated RemoteClusterAuthorizationResolver to return the encoded key with the prefix, removed EncodedApiKeyWithPrefix, and added a call to exercise API key parsing inside readFromContext to force validation. This achieves our goals above.